tracee-ebpf/tracee/events_pipeline.go

```diff
@@ -84,69 +99,79 @@ func (t *Tracee) processEvents(done <-chan struct{}) error {
 		// Todo: The monotonic clock doesn't take into account system sleep time.
 		// Starting from kernel 5.7, we can get the timestamp relative to the system boot time instead which is preferable.
 		if t.config.Output.RelativeTime {
 			// To get the monotonic time since tracee was started, we have to substract the start time from the timestamp.
-			ctx.Ts -= t.startTime
+			eventCtx.Ts -= t.startTime
 		} else {
 			// To get the current ("wall") time, we add the boot time into it.
-			ctx.Ts += t.bootTime
+			eventCtx.Ts += t.bootTime
 		}
 
 		evt := external.Event{
-			Timestamp:           int(ctx.Ts),
-			ProcessID:           int(ctx.Pid),
-			ThreadID:            int(ctx.Tid),
-			ParentProcessID:     int(ctx.Ppid),
-			HostProcessID:       int(ctx.HostPid),
-			HostThreadID:        int(ctx.HostTid),
-			HostParentProcessID: int(ctx.HostPpid),
-			UserID:              int(ctx.Uid),
-			MountNS:             int(ctx.MntID),
-			PIDNS:               int(ctx.PidID),
-			ProcessName:         string(bytes.TrimRight(ctx.Comm[:], "\x00")),
-			HostName:            string(bytes.TrimRight(ctx.UtsName[:], "\x00")),
+			Timestamp:           int(eventCtx.Ts),
+			ProcessID:           int(eventCtx.Pid),
+			ThreadID:            int(eventCtx.Tid),
+			ParentProcessID:     int(eventCtx.Ppid),
+			HostProcessID:       int(eventCtx.HostPid),
+			HostThreadID:        int(eventCtx.HostTid),
+			HostParentProcessID: int(eventCtx.HostPpid),
+			UserID:              int(eventCtx.Uid),
+			MountNS:             int(eventCtx.MntID),
+			PIDNS:               int(eventCtx.PidID),
+			ProcessName:         string(bytes.TrimRight(eventCtx.Comm[:], "\x00")),
+			HostName:            string(bytes.TrimRight(eventCtx.UtsName[:], "\x00")),
 			ContainerID:         containerId,
-			EventID:             int(ctx.EventID),
+			EventID:             int(eventCtx.EventID),
 			EventName:           eventDefinition.Name,
-			ArgsNum:             int(ctx.Argnum),
-			ReturnValue:         int(ctx.Retval),
+			ArgsNum:             int(eventCtx.Argnum),
+			ReturnValue:         int(eventCtx.Retval),
 			Args:                args,
 			StackAddresses:      StackAddresses,
 		}
 
-		err = t.processEvent(&evt)
+		select {
+		case <-ctx.Done():
+			return
+		case out <- &evt:
+			t.stats.eventCounter.Increment()
+		}
+	}
+}
+
+func (t *Tracee) processEvents(ctx context.Context, in <-chan *external.Event) {
+	for event := range in {
+		err := t.processEvent(event)
 		if err != nil {
 			t.handleError(err)
 			continue
 		}
 
 		// Only emit events requested by the user
-		if !t.eventsToTrace[ctx.EventID] {
+		if !t.eventsToTrace[int32(event.EventID)] {
 			continue
 		}
 
 		if t.config.Output.ParseArguments {
-			err = t.parseArgs(&evt)
+			err = t.parseArgs(event)
 			if err != nil {
 				t.handleError(err)
 				continue
 			}
 		}
 
 		select {
-		case <-done:
-			return nil
-		case t.config.ChanEvents <- evt:
+		case <-ctx.Done():
+			return
+		case t.config.ChanEvents <- event:
```

Comment:
why can't we just write `*event` here and preserve the external interface?

Reply:
I think this should happen in the pipeline, no in the pipeline step. it makes the pipeline less readable (you look at the handleevent and you don't know what's going to happen), and less testable

Reply:
Using *event instead of event will do a copy of the event struct, which is not required.
By moving the pointer to the created event, we can avoid performing such copies, and work on a single instance of the event, which is preferrable.

Reply:
> I think this should happen in the pipeline, no in the pipeline step. it makes the pipeline less readable (you look at the handleevent and you don't know what's going to happen), and less testable

This is the expected behavior of pipelines in go. A pipline has a source stage and a sink stage, as implemented here. I don't think we should implement it differently (https://go.dev/blog/pipelines)

Reply:
> A pipline has a source stage and a sink stage, as implemented here

yes, a pipeline stage should have a source and a sink, but this code doesn't. What is the source and the sink here?

> I don't think we should implement it differently (https://go.dev/blog/pipelines)

I agree we should implement it like in that post but the code here is different

Reply:
> > A pipline has a source stage and a sink stage, as implemented here
> 
> yes, a pipeline stage should have a source and a sink, but this code doesn't. What is the source and the sink here?

As we only have two stages, this should be clear - the source is the decode events stage and the sink is the process events stage. You can argue that these stages have input and output channels, but these are not part of the pipeline. The input channel for decode events is for getting the events from the bpf code and the output channel is what has been configured for the events channel by the consumer of tracee-ebpf.

> 
> > I don't think we should implement it differently (https://go.dev/blog/pipelines)
> 
> I agree we should implement it like in that post but the code here is different

It is different by two things that I changed with the following reasons:
1. We don't return `outc`, `errc` and `err`: As we already have `handleError()` function which sends errors through a dedicated error channel, I didn't see a reason to add an extra error channel per each stage. Also, we never return an error which terminates the for loop, so there is no reason for having `err`. As we don't need errc and err, I decided to create the output channel outside the pipeline stages, which brings me to the second difference.
2. We don't call the `go` keyword from within the pipeline stage, but can call it from the outside, like I did here (which also avoids an extra indentation, and makes the PR easier to review).

Reply:
> We don't return outc, errc and err: As we already have handleError() function which sends errors through a dedicated error channel, I didn't see a reason to add an extra error channel per each stage.

In this matter I think the `handleError()` method is a problem, because it bounds you to create all go routines to be methods of `tracee`.
In my opinion, we should add an error channel which leads to a go routine that prints the errors and update the metrics. This way the new stages of the pipeline could be in different packages.

Reply:
> > We don't return outc, errc and err: As we already have handleError() function which sends errors through a dedicated error channel, I didn't see a reason to add an extra error channel per each stage.
> 
> In this matter I think the `handleError()` method is a problem, because it bounds you to create all go routines to be methods of `tracee`. In my opinion, we should add an error channel which leads to a go routine that prints the errors and updates the metrics (in other words calling to `handleError` itself). This way the new stages of the pipeline could be in different packages.

Yes, you are right. We should handle errors in a different way so we can decouple the different stages. Currently, handleError is also called from net_events.go and write_capture.go, and I'm not going to handle this decoupling in this PR. I will update my PR though to include error channel and merging errors, like we had in the past.

Reply:
> It is different by two things that I changed with the following reasons: We don't return outc, errc and err

That's what I'm trying to say. this implementation is not the pipeline that's described in the blog post. it's big difference since by not returning out and not receiving in you are coupling the stages (also with shared state)

Bottom line is that a proper pipeline should control the flow of execution, and the pipeline stages should not be involved. I don't think it's a huge deal but I think it would be changed to a proper pipeline at some point probably (it it would remain a pipeline)

Reply:
> > It is different by two things that I changed with the following reasons: We don't return outc, errc and err
> 
> That's what I'm trying to say. this implementation is not the pipeline that's described in the blog post. it's big difference since by not returning out and not receiving in you are coupling the stages (also with shared state)
> 
> Bottom line is that a proper pipeline should control the flow of execution, and the pipeline stages should not be involved. I don't think it's a huge deal but I think it would be changed to a proper pipeline at some point probably (it it would remain a pipeline)

Technically speaking, there is no real difference between what I implemented and what described in this blog. But let's not continue this discussion - have you seen the recent changes? I don't think so, because now it is implemented in the exact same way.

Reply:
> > > why can't we just write *event here and preserve the external interface?
> 
> > Using *event instead of event will do a copy of the event struct, which is not required.
> > By moving the pointer to the created event, we can avoid performing such copies, and work on a single instance of the event, which is preferrable.
> 
> I don't think it matters if we do another copy or not (happy to be proven wrong here). As long as it's internal code then I wouldn't care if you wanted to share pointers but when it comes out of tracee, I think the API should remain a channel of struct and not a channel of pointer to struct.

Why?